### ISSUE TITLE
twisted.names.dns.Messages should be compared based on their field values and section contents

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6848]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6848|
|Type|enhancement|
|Created|2013-11-21 17:48:45Z|
|Branch|https://github.com/twisted/twisted/tree/rich-message-comparison-6848|

Some twisted.names tests would be much easier to write if dns.Message implemented some of the rich comparison methods and if the Messages were compared based on their field values and section contents.

The simplest way to do this is to inherit from FancyEqMixin.

Which is what we've don in the proposed EDNSMessage class in [#5675](https://github.com/twisted/twisted/issues/5675)

See https://twistedmatrix.com/trac/browser/branches/edns-message-5675-4/twisted/names/dns.py?rev=40240#L2228

And see [#6847](https://github.com/twisted/twisted/issues/6847) where I've added FancyStrMixin for a nicer Message repr.

<details><summary>Searchable metadata</summary>

```
trac-id__6848 6848
type__enhancement enhancement
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_rich_message_comparison_6848 branches/rich-message-comparison-6848
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1385056125000000 1385056125000000
changetime__1386420964000000 1386420964000000
version__None None
owner__tom_prince tom.prince

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_1"></a>|@wallrj commented|
|-|-|

(In [40844]) Branching to 'rich-message-comparison-6848'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_2"></a>|@wallrj commented|
|-|-|

Ready for review in log:branches/rich-message-comparison-6848
 * Added FancyEqMixin and associated tests
 * This and [#6847](https://github.com/twisted/twisted/issues/6847) should allow some shared rich comparison tests between dns.Message and dns.EDNSMessage in [#5675](https://github.com/twisted/twisted/issues/5675)
 * Not sure how adding rich comparison and repr fits in with the compatibility policy. 
 * It would be nice if showAttributes and compareAttributes could be private, that'll require a change to FancyEq and FancyStrMixin. Good idea?

Build Results:
 * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/rich-message-comparison-6848

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince set owner to @tomprince|
|-|-|

This looks good. I'll merge it. I reordered the attributes and tests to match the constructor argument order, to make it easier to review.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_4"></a>|@tomprince set status to `closed`|
|-|-|

(In [40930]) Merge rich-message-comparison-6848: Compare twisted.names.dns.Messages based on its attributes.

Author: rwall
Reviewers: tom.prince
Fixes: [#6848](https://github.com/twisted/twisted/issues/6848)

Some twisted.names tests will be much easier to write if dns.Message
has a value based equality, rather than object identity based equality.
